examples/src/ExampleInstrumentation.php

```diff
@@ -0,0 +1,55 @@
+<?php
+
+declare(strict_types=1);
+
+namespace OpenTelemetry\Example;
+
+use Exception;
+use OpenTelemetry\API\Globals;
+use OpenTelemetry\API\Instrumentation\AutoInstrumentation\ConfigurationRegistry;
+use OpenTelemetry\API\Instrumentation\AutoInstrumentation\HookManager;
+use OpenTelemetry\API\Instrumentation\AutoInstrumentation\Instrumentation;
+use OpenTelemetry\API\Trace\Span;
+use OpenTelemetry\Context\Context;
+
+final class ExampleInstrumentation implements Instrumentation
+{
+
+    /**
+     * @todo can we pass in just the config for _this_ instrumentation, rather than all?
+     */
+    public function register(HookManager $hookManager, ConfigurationRegistry $configuration): void
+    {
+        $config = $configuration->get(ExampleConfig::class) ?? throw new Exception('example instrumentation must be configured');
+        if (!$config->enabled) {
+            return;
+        }
+
+        $tracer = Globals::tracerProvider()->getTracer('example-instrumentation');
```

Comment:
This triggers initialization during sdk autoloading; iff we want to use the `Globals` providers to delay initialization, we should specify that `Globals` must not be accessed outside of hook callbacks. Otherwise I don't see an advantage over injecting the providers.

Apart from that I would generally prefer if we inject the providers. It allows initializing telemetry instances outside of hook callbacks¹ and might allow us to change `Globals` to be a truly global state instead of being `Context` dependent.

¹ For example reusing a counter instance and calling `::add()` directly is around three times faster than refetching the internal instrument from a `CachedInstrumentation` instance. (initial counter creation w/ single metric reader/metric stream takes ~33μs):

```
+---------------------------------+---------+----------+--------+---------+
| subject                         | memory  | mode     | rstdev | stdev   |
+---------------------------------+---------+----------+--------+---------+
| benchCounterAdd (empty)         | 2.716mb | 0.902μs  | ±1.95% | 0.018μs |
| benchCounterAddRecreate (empty) | 2.780mb | 3.319μs  | ±1.86% | 0.062μs |
| benchCounterAdd (http)          | 2.727mb | 1.954μs  | ±1.63% | 0.032μs |
| benchCounterAddRecreate (http)  | 2.783mb | 4.539μs  | ±1.73% | 0.079μs |
+---------------------------------+---------+----------+--------+---------+
```

Reply:
That makes sense, and I prefer the method of passing in the providers when registering an instrumentation. I don't think we can commit to file-based config and deprecate `Registry` just yet, given the experimental state of config.

But, I think we can start to move in that direction. How does this sound?
1. update existing instrumentation classes to only retrieve providers from hook callbacks
2. document this new restriction (probably in `Globals`)
3. add a todo to Registry to deprecate in a future release (in case anybody's reading the source)
4. pass in Noop providers to `Instrumentation`s via `::register()`, and document that these will become real implementations in a future release but to keep using `Globals` for now

We then wait for config to stabilise.

Reply:
> I don't think we can commit to file-based config and deprecate Registry just yet, given the experimental state of config.

The restriction / issue with `Registry` is not related to file-based config; any usage of `Globals` within an `autoload.files` file might trigger this issue already, for example [`open-telemetry/opentelemetry-auto-openai-php`](https://github.com/open-telemetry/opentelemetry-php-contrib/blob/f9b3ca8c83c74a19b62ce0bf96ba8dacf19ca1a7/src/Instrumentation/OpenAIPHP/src/OpenAIPHPInstrumentation.php#L45-L49).

4. If we have to stay compatible with `Registry::register{Xxx}Factory` for now:
Instead of noop providers we could inject providers that resolve `Globals` on the first call to `::spanBuilder()` / `::create{Instrument}()` / `::emit()`. This would allow us to use the injected providers immediately (with the same restriction as `CachedInstrumentation` that `Tracer`/`Meter`/`Logger` methods must only be called within hook callbacks for now).

<details>
<summary>TracerProvider example</summary>

```php
$tracerProvider = new LateBindingTracerProvider(static function(): TracerProviderInterface {
    $scope = Context::getRoot()->activate();
    try {
        return Globals::tracerProvider();
    } finally {
        $scope->detach();
    }
});

final class LateBindingTracerProvider implements TracerProviderInterface {

    private ?TracerProviderInterface $tracerProvider = null;

    /** @param Closure(): TracerProviderInterface $factory */
    public function __construct(
        private readonly Closure $factory,
    ) {}

    public function getTracer(string $name, ?string $version = null, ?string $schemaUrl = null, iterable $attributes = []): TracerInterface {
        return $this->tracerProvider?->getTracer($name, $version, $schemaUrl, $attributes)
            ?? new LateBindingTracer(fn(): TracerInterface => ($this->tracerProvider ??= ($this->factory)())->getTracer($name, $version, $schemaUrl, $attributes));
    }
}

final class LateBindingTracer implements TracerInterface {

    private ?TracerInterface $tracer = null;

    /** @param Closure(): TracerInterface $factory */
    public function __construct(
        private readonly Closure $factory,
    ) {}

    public function spanBuilder(string $spanName): SpanBuilderInterface {
        return ($this->tracer ??= ($this->factory)())->spanBuilder($spanName);
    }
}
```
</details>

Reply:
Implemented late-binding providers based on the above example (thanks!)